src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncSpillTests.cs

```diff
@@ -3888,5 +3887,2228 @@ class AltBoolean
+    IL_0074:  stloc.1
+    IL_0075:  ldarg.0
+    IL_0076:  ldfld      ""A Program.<Assign>d__0.<>7__wrap1""
+    IL_007b:  ldloc.1
+    IL_007c:  stfld      ""int A.x""
+    IL_0081:  ldarg.0
+    IL_0082:  ldnull
+    IL_0083:  stfld      ""A Program.<Assign>d__0.<>7__wrap1""
+    IL_0088:  leave.s    IL_00a1
+  }
+  catch System.Exception
+  {
+    IL_008a:  stloc.3
+    IL_008b:  ldarg.0
+    IL_008c:  ldc.i4.s   -2
+    IL_008e:  stfld      ""int Program.<Assign>d__0.<>1__state""
+    IL_0093:  ldarg.0
+    IL_0094:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder Program.<Assign>d__0.<>t__builder""
+    IL_0099:  ldloc.3
+    IL_009a:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetException(System.Exception)""
+    IL_009f:  leave.s    IL_00b4
+  }
+  IL_00a1:  ldarg.0
+  IL_00a2:  ldc.i4.s   -2
+  IL_00a4:  stfld      ""int Program.<Assign>d__0.<>1__state""
+  IL_00a9:  ldarg.0
+  IL_00aa:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder Program.<Assign>d__0.<>t__builder""
+  IL_00af:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
+  IL_00b4:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(42755, "https://github.com/dotnet/roslyn/issues/42755")]
+        public void KeepLtrSemantics_StructFieldAccessOnArray()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program
+{
+    static async Task Assign(A[] arr)
+    {
+        arr[0].x = await Write(""RHS"");
+    }
+
+    static async Task Main(string[] args)
+    {
+        await TestIndexerThrows();
+        await TestIndexerSucceeds();
+        await TestReassignsArrayAndIndexerDuringAwait();
+        await TestReassignsTargetDuringAwait();
+    }
+
+    static async Task TestIndexerThrows()
+    {
+        Console.WriteLine(nameof(TestIndexerThrows));
+        
+        var arr = new A[0];
+        Console.WriteLine(""Before Assignment"");
+        try
+        {
+            await Assign(arr);
+        }
+        catch (IndexOutOfRangeException)
+        {
+            Console.WriteLine(""Caught IndexOutOfRangeException"");
+        }
+    }
+
+    static async Task TestIndexerSucceeds()
+    {
+        Console.WriteLine(nameof(TestIndexerSucceeds));
+
+        var arr = new A[1];
+        Console.WriteLine(""Before Assignment arr[0].x is: "" + arr[0].x);
+        await Assign(arr);
+        Console.WriteLine(""After Assignment arr[0].x is: "" + arr[0].x);
+    }
+
+    static async Task TestReassignsArrayAndIndexerDuringAwait()
+    {
+        Console.WriteLine(nameof(TestReassignsArrayAndIndexerDuringAwait));
+
+        var arr = new A[1];
+        var index = 0;
+        Console.WriteLine(""Before Assignment arr.Length is: "" + arr.Length);
+        arr[index].x = await WriteAndReassign(""RHS"");
+        Console.WriteLine(""After Assignment arr.Length is: "" + arr.Length);
```

Comment:
>Console.WriteLine(""After Assignment arr.Length is: "" + arr.Length); [](start = 8, length = 69)

In addition should assert that the expected element in the original array is modified as we expect.